content/learning-paths/cross-platform/loop-reflowing/autovectorization-and-restrict.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Autovectorization and restrict
+weight: 3
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+## Autovectorization and restrict keyword
+
+You have already experienced some form of autovectorization by learning about the [`restrict` keyword in a previous Learning Path](https://learn.arm.com/learning-paths/cross-platform/restrict-keyword-c99/).
+Our example is a classic textbook example that the compiler will autovectorize simply by using `restrict`:
+
+Try the previously saved files, compile them both and compare the assembly output:
+
+```bash
+gcc -O2 addvec.c -o addvec
+gcc -O2 addvec_neon.c -o addvec_neon
```

Comment:
ok, will change that to -O1.

Reply:
I just tested that, I just tested clang and gcc on this loop on godbolt: https://godbolt.org/z/7v5qsEsP5
As you already noted, clang -O2 autovectorizes the loop, but gcc -O2 doesn't, however if I add restrict to C then gcc also autovectorizes the loop. But you're saying that this might change in the future for gcc?

Reply:
yes, the current decision for GCC to vectorize or not is driven by the cost model, at `-O2` GCC uses the `very cheap` cost model, which means that vectorization must be a clear win. The reason we don't vectorize this loop is that this cost model determines that if we emit the alias check we grow code size, and at `-O2` codesize is considered more.  That's why the restrict works, but there are other ways the compiler can prove loops do not alias. e.g. type based aliasing https://godbolt.org/z/zsYhYrKeY etc. So the decision to vectorize at `-O2` is determined by many things, which will likely shift over time.

Reply:
Thank you for the explanation, I will include this as a note!